File: 2 - Work/Phase 5 The Hawaii Micro-Case Study/Bulk Tests/python/Step2_Parcel_Intersection.py
# Purpose: Intersect Oahu OSM golf polygons with Honolulu parcel cadastre to
#          extract TMK identifiers and total legal footprint area.
# Inputs:  Bulk Tests/python/Target_Golf_Polygons.gpkg        (Step 1 output)
#          Bulk Tests/python/Honolulu_Parcels_Reprojected.gpkg (Step 1 output)
# Outputs: Bulk Tests/python/Target_Golf_Parcels_List.csv


# === 1. IMPORTS ===

import pathlib
import geopandas as gpd
import pandas as pd


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR       = pathlib.Path(__file__).parent
TARGET_GOLF_PATH = SCRIPT_DIR / "Target_Golf_Polygons.gpkg"
PARCELS_PATH     = SCRIPT_DIR / "Honolulu_Parcels_Reprojected.gpkg"
OUT_CSV          = SCRIPT_DIR / "Target_Golf_Parcels_List.csv"

TMK_CANDIDATES = [
    "TMK", "PARCEL_ID", "Parcel_ID", "parcel_id", "TAX_MAP_KEY",
    "Tax_Map_Key", "tax_map_key", "MAPKEY", "mapkey", "tmk",
]


# === 3. EXECUTION ===

def main():
    if not TARGET_GOLF_PATH.exists():
        print("[FATAL] Target Golf Polygons not found. Run Step 1.")
        raise SystemExit(1)
    if not PARCELS_PATH.exists():
        print("[FATAL] Reprojected Parcels not found. Run Step 1.")
        raise SystemExit(1)

    print("Phase 5 - Step 2: Parcel Intersection")
    print(f"Loading datasets from: {SCRIPT_DIR}")

    # [METHODOLOGY] gpd.read_file — spatial read of Step 1 OSM golf polygons
    target_golf_geo = gpd.read_file(TARGET_GOLF_PATH)
    # [METHODOLOGY] gpd.read_file — spatial read of Step 1 reprojected parcel cadastre
    parcels_geo     = gpd.read_file(PARCELS_PATH)

    print(f"  -> Loaded {len(target_golf_geo)} target golf polygons.")
    print(f"  -> Loaded {len(parcels_geo)} parcel features.")

    print("\nPerforming spatial intersection (this may take a moment)...")
    # [METHODOLOGY] gpd.overlay(how='intersection') — cookie-cutter of Phase 2 OSM polygons
    #               over the Phase 5 legal cadastre to isolate golf-course parcel fragments
    parcel_intersection_geo = gpd.overlay(target_golf_geo, parcels_geo, how="intersection")

    print(f"  -> Intersection complete: {len(parcel_intersection_geo)} parcel fragments found.")

    print("\nExtracting unique TMK identifiers...")

    found_column = None
    for col in TMK_CANDIDATES:
        if col in parcel_intersection_geo.columns:
            found_column = col
            break

    if found_column is None:
        print("\n[WARNING] Standard TMK column not found. Available columns:")
        print(list(parcel_intersection_geo.columns))
        print("[FATAL] No TMK column identified.")
        raise SystemExit(1)

    raw_tmks          = parcel_intersection_geo[found_column].dropna().astype(str).unique()
    unique_tmk_sorted = sorted(raw_tmks)

    print(f"  -> Found {len(unique_tmk_sorted)} unique TMKs across the {len(target_golf_geo)} golf courses.")

    tmk_df = pd.DataFrame({"TMK": unique_tmk_sorted})
    tmk_df.to_csv(OUT_CSV, index=False)

    # [METHODOLOGY] .geometry.area — compute legal footprint area from intersection geometry
    total_area_m2   = parcel_intersection_geo.geometry.area.sum()
    total_acres     = total_area_m2 / 4046.86
    formatted_acres = f"{total_acres:,.2f}"

    print("\n" + "=" * 60)
    print("PARCEL INTERSECTION COMPLETE")
    print("=" * 60)
    print(f"  Total Targeted Courses : {len(target_golf_geo)}")
    print(f"  Total Unique TMKs      : {len(unique_tmk_sorted)}")
    print(f"  Total Legal Footprint  : {formatted_acres} Acres")
    print("-" * 60)
    print(f"[+] Exported TMK List (CSV) : {OUT_CSV.as_posix()}")
    print("\n[DONE] Step 2 Complete.")


if __name__ == "__main__":
    main()

File: 2 - Work/Phase 5 The Hawaii Micro-Case Study/Bulk Tests/R/Step2_Parcel_Intersection.R
# Purpose: Intersect Oahu OSM golf polygons with Honolulu parcel cadastre to
#          extract TMK identifiers and total legal footprint area.
# Inputs:  Bulk Tests/R/Target_Golf_Polygons.gpkg        (Step 1 output)
#          Bulk Tests/R/Honolulu_Parcels_Reprojected.gpkg (Step 1 output)
# Outputs: Bulk Tests/R/Target_Golf_Parcels_List.csv


# === 1. LIBRARIES ===

suppressPackageStartupMessages({
    library(sf)
    library(tidyverse)
    library(this.path)
})


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR       <- this.path::this.dir()
TARGET_GOLF_PATH <- file.path(SCRIPT_DIR, "Target_Golf_Polygons.gpkg")
PARCELS_PATH     <- file.path(SCRIPT_DIR, "Honolulu_Parcels_Reprojected.gpkg")
OUT_CSV          <- file.path(SCRIPT_DIR, "Target_Golf_Parcels_List.csv")


# === 3. EXECUTION ===

cat("Phase 5b - Step 2: Parcel Intersection\n")
cat(sprintf("Loading datasets from: %s\n", SCRIPT_DIR))

if (!file.exists(TARGET_GOLF_PATH)) stop("[FATAL] Target Golf Polygons not found. Run Step 1.")
if (!file.exists(PARCELS_PATH))     stop("[FATAL] Reprojected Parcels not found. Run Step 1.")

# [METHODOLOGY] st_read — spatial read of Step 1 OSM golf polygons
target_golf_sf <- st_read(TARGET_GOLF_PATH, quiet = TRUE)
# [METHODOLOGY] st_read — spatial read of Step 1 reprojected parcel cadastre
parcels_sf     <- st_read(PARCELS_PATH, quiet = TRUE)

cat(sprintf("  -> Loaded %d target golf polygons.\n", nrow(target_golf_sf)))
cat(sprintf("  -> Loaded %d parcel features.\n", nrow(parcels_sf)))

cat("\nPerforming spatial intersection (this may take a moment)...\n")
# [METHODOLOGY] st_intersection — cookie-cutter of Phase 2 OSM polygons over the
#               Phase 5 legal cadastre to isolate golf-course parcel fragments
parcel_intersection_sf <- st_intersection(target_golf_sf, parcels_sf)

cat(sprintf(
    "  -> Intersection complete: %d parcel fragments found.\n",
    nrow(parcel_intersection_sf)
))

cat("\nExtracting unique TMK identifiers...\n")

tmk_columns <- c(
    "TMK", "PARCEL_ID", "Parcel_ID", "parcel_id", "TAX_MAP_KEY",
    "Tax_Map_Key", "tax_map_key", "MAPKEY", "mapkey", "tmk"
)

found_column <- NULL
for (col in tmk_columns) {
    if (col %in% names(parcel_intersection_sf)) {
        found_column <- col
        break
    }
}

if (is.null(found_column)) {
    cat("\n[WARNING] Standard TMK column not found. Available columns:\n")
    print(names(parcel_intersection_sf))
    stop("[FATAL] No TMK column identified.")
}

unique_tmk        <- unique(as.character(parcel_intersection_sf[[found_column]]))
unique_tmk_sorted <- sort(unique_tmk[!is.na(unique_tmk)])

cat(sprintf(
    "  -> Found %d unique TMKs across the %d golf courses.\n",
    length(unique_tmk_sorted), nrow(target_golf_sf)
))

tmk_df <- data.frame(TMK = unique_tmk_sorted)
write.csv(tmk_df, OUT_CSV, row.names = FALSE)

# [METHODOLOGY] st_area — compute legal footprint area from intersection geometry
total_area_m2   <- sum(st_area(parcel_intersection_sf))
total_acres     <- as.numeric(total_area_m2) / 4046.86
formatted_acres <- formatC(total_acres, format = "f", big.mark = ",", digits = 2)

cat("\n============================================================\n")
cat("PARCEL INTERSECTION COMPLETE\n")
cat("============================================================\n")
cat(sprintf("  Total Targeted Courses : %d\n", nrow(target_golf_sf)))
cat(sprintf("  Total Unique TMKs      : %d\n", length(unique_tmk_sorted)))
cat(sprintf("  Total Legal Footprint  : %s Acres\n", formatted_acres))
cat("------------------------------------------------------------\n")
cat(sprintf(
    "[+] Exported TMK List (CSV) : %s\n",
    normalizePath(OUT_CSV, winslash = "/")
))
cat("\n[DONE] Step 2 Complete.\n")
